plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/EncoderServer.kt

```diff
@@ -0,0 +1,260 @@
+import software.aws.toolkits.core.utils.exists
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.core.utils.tryDirOp
+import software.aws.toolkits.jetbrains.services.codewhisperer.settings.CodeWhispererSettings
+import software.aws.toolkits.jetbrains.services.cwc.editor.context.project.manifest.ManifestManager
+import java.io.FileOutputStream
+import java.io.IOException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.nio.file.attribute.PosixFilePermission
+import java.security.Key
+import java.security.SecureRandom
+import java.util.Base64
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.zip.ZipFile
+import javax.crypto.spec.SecretKeySpec
+
+class EncoderServer(val project: Project) : Disposable {
+    private val cachePath = Paths.get(
+        UserHomeDirectoryUtils.userHomeDirectory()
+    ).resolve(".aws").resolve("amazonq").resolve("cache")
+    val manifestManager = ManifestManager()
+    private val serverDirectoryName = "qserver-${manifestManager.currentVersion}.zip"
+    private var numberOfRetry = AtomicInteger(0)
+    val port by lazy { NetUtils.findAvailableSocketPort() }
+    private val nodeRunnableName = if (manifestManager.getOs() == "windows") "node.exe" else "node"
+    private val maxRetry: Int = 3
+    val key = generateHmacKey()
+    private var processHandler: KillableProcessHandler? = null
+
+    fun downloadArtifactsAndStartServer() {
+        downloadArtifactsIfNeeded()
+        start()
+    }
+
+    fun isNodeProcessRunning() = processHandler != null && processHandler?.process?.isAlive == true
+
+    private fun generateHmacKey(): Key {
+        val keyBytes = ByteArray(32)
+        SecureRandom().nextBytes(keyBytes)
+        return SecretKeySpec(keyBytes, "HmacSHA256")
+    }
+
+    fun encrypt(data: String): String {
+        val header = JWSHeader.Builder(JWSAlgorithm.HS256)
+            .type(JOSEObjectType.JWT)
+            .build()
+
+        val claimsSet = JWTClaimsSet.Builder()
+            .subject(Base64.getUrlEncoder().withoutPadding().encodeToString(data.toByteArray()))
+            .build()
+
+        val signedJWT = SignedJWT(header, claimsSet)
+        signedJWT.sign(MACSigner(key.encoded))
+
+        return signedJWT.serialize()
+    }
+
+    data class EncryptionRequest(
+        val version: String = "1.0",
+        val mode: String = "JWT",
+        val key: String,
+    )
+
+    fun getEncryptionRequest(): String {
+        val request = EncryptionRequest(key = Base64.getUrlEncoder().withoutPadding().encodeToString(key.encoded))
+        return jacksonObjectMapper().writeValueAsString(request)
+    }
+
+    private fun runCommand(command: GeneralCommandLine): Boolean {
+        try {
+            logger.info("starting encoder server for project context on $port for ${project.name}")
+            processHandler = KillableProcessHandler(command)
+            val exitCode = processHandler!!.waitFor()
```

Comment:
maybe make it a lateinit var instead of NULL as default value 
```suggestion
           private lateinit var  processHandler: KillableProcessHandler
```

Reply:
since we try to destroy the process when there is exception I feel it's safer to make it a nullable